# Afinação e Notas Musicais

## Objetivo
Após esta unidade, o aluno será capaz de aplicar modelos matemáticos para relacionar o fenômeno perceptual da **altura**, o fenômeno físico da **frequência fundamental** e o fenômeno cultural das **notas musicais**.

## Pré-requisitos
Para acompanhar adequadamente esta unidade, o aluno deve estar tranquilo com:
1. Podemos organizar notas musicais em oitavas,
1. Os nomes das notas musicais se repetem em cada oitava,
1. Algumas notas são referidas usando um modificador sustenido ou bemol
1. Quando duas notas soam juntas, o som resultante pode ser dissonante ou consonante.

## Como afinar uma lira

O conceito de afinação está bastante ligado ao conceito de *intervalos*. Um intervalo é a diferença perceptual de alturas entre dois tons que soam simultaneamente. Tons medidos e absolutos (por exemplo: uma senóide de frequência 440 Hz) só foram possíveis depois da invenção de instrumentos modernos de geração e de instrumentos precisos para medição. Antes destes avanços tecnológicos, só era possível afinar instrumentos utilizando a percepção de intervalos.

Na Grécia antiga, já haviam instrumentos de sopro e de corda (veja a flauta de Baco ou a lira de Apolo, por exemplo). Isso significa que havia alguma forma de afiná-los (em especial, o instrumento de corda, que desafina rapidamente com o calor). Portanto, os gregos já dominavam o conceito de que uma corda emite um som mais agudo quando é esticada mais fortemente, e um som mais grave quando é esticada com menos força.

### Afinação Pitagórica
[Pitágoras](http://www.sonora.iar.unicamp.br/index.php/sonora1/article/viewFile/29/28) é dito ser um dos pioneiros em sistematizar uma forma de afinar instrumentos. Ele observou que duas cordas de igual material, esticadas sob a mesma força, mas de comprimentos diferentes, [emitem sons de alturas diferentes](https://courses.physics.illinois.edu/phys193/student_reports/fall06/j_wilson/j_wilson_p199pom_fa06_final_pres.pdf). Sabemos, hoje, que cada altura diferente está ligada a uma frequência fundamental diferente.

A união dos sons simultâneos das duas cordas produz um *intervalo*. Pitágoras observou que o intervalo produzido era muito mais agradável ao ouvido (ou: consonante) quando a razão entre os comprimentos das cordas era de 1 para 2. Neste caso, sabemos que a frequência fundamental de uma das vibrações é a metade da outra, e este intervalo é chamado de *oitava*. O segundo intervalo mais agradável ao ouvido ocorria quando os comprimentos das cordas tinham razão 2 para 3. Neste caso, surge um intervalo chamado de *quinta*.

Pitágoras, então, definiu o seguinte método para encontrar as alturas de uma escala:
1. Inicie com um tom de referência
1. Afine a próxima nota usando um intervalo de quinta na direção dos agudos
1. Se meu intervalo em relação à referência é maior que uma oitava, então caminhe uma oitava em direção dos graves
1. Se já afinei todas as cordas de uma oitava, então pare.
1. Use a nova nota obtida como referência e continue do passo 2.

In [14]:
referencia_inicial = 440.0 # Hz
frequencias = [] # Esta lista recebera todas as frequencias de uma escala

f = referencia_inicial
while len(frequencias) < 12:
    if f > (referencia_inicial * 2):
        f /= 2.
    frequencias.append(f)
    f *= (3/2.)


frequencias.sort()
print frequencias
print f

[440.0, 469.86328125, 495.0, 528.59619140625, 556.875, 594.6707153320312, 626.484375, 660.0, 704.794921875, 742.5, 792.894287109375, 835.3125]
892.006072998


Veja que há um fenômeno interessante que acontece. A nota que deveria ter frequência de 880.0 Hz (um intervalo de oitava em relação à referência inicial) na verdade é calculada como 892 Hz. Isso representa um erro calculável, na forma:

In [10]:
print 100*(f - (referencia_inicial * 2)) / (referencia_inicial*2)
       

1.36432647705


Esse erro, de 1.36%, é bem conhecido e é chamado de Coma Pitagórico. Ele representa uma espécie de erro perceptual acumulado do sistema de afinação. Mas, veja: há um erro perceptual acumulado que, causa uma sensação de dissonância mesmo que a afinação toda utilize apenas intervalos perfeitamente consonantes. Trata-se de um paradoxo com o qual os músicos tiveram que lidar através da história.

Na verdade, isso é um [fenômeno matematicamente inevitável](http://legacy.earlham.edu/~tobeyfo/musictheory/Book1/FFH1_CH1/1N_Pythagorean_Comma.html). A série de frequências geradas pela afinação pitagórica tem a forma:
$$f \times (\frac{3}{2}) ^ N,$$
excluídas as oitavas de ajuste.

Um intervalo de oitava é gerado utilizando uma multiplicação da frequência inicial por uma potência de dois e, como sabemos, não existe nenhum caso em que uma potência de dois é gerada por uma potência de três.

### Nomes das notas musicais
Por razões históricas e puramente culturais, as civilizações que descenderam da Grécia antiga utilizaram doze notas como as partes de uma escala. Poderiam ser mais (como em algumas culturas asiáticas) ou menos (como em algumas culturas africanas). As notas receberam seus nomes (dó, ré, mi, fá, sol, lá, si) tomando por base uma [poesia em latim](https://en.wikipedia.org/wiki/Ut_queant_laxis) chamada Ut Queant Laxis, que é um hino a João Batista. Também, é comum utilizar a notação em letras (C, D, E, F, G, A, B) e os sinais de bemol (Cb, Db, Eb, etc) e sustenido (C#, D#, E#, etc) para denotar acidentes.


### Ciclo de Quintas
O sistema de afinação Pitagórico determina, implicitamente, um caminho por entre as notas musicais de uma escala que foi historicamente chamado de *ciclo de quintas*. Trata-se de um ciclo onde se colocam todas as notas musicais (excluindo-se a oitava), de forma que um passo em qualquer direção percorre um intervalo de quinta:

C, G, D, A, E, B, F#, C#, G#, D#, A#, F, C

### Afinação de igual temperamento
A relação entre a frequência de vibração de uma corda e a sensação de altura relacionada a ela foi estudada incialmente (ao menos nestes termos) por Vincenzo Galilei, no século XVI. Esse conhecimento permitiu usar uma linguagem mais próxima à contemporânea para se referir à frequência de cordas vibrantes. Mesmo assim, ainda não haviam instrumentos para detectar frequências com precisão, e, portanto, os sistemas de afinação ainda dependiam de relações perceptuais intervalares.

Uma afinação relativa, como a pitagórica, varia de acordo com o tom. Isso significa que, se um cravo for afinado à partir de um Dó, ele será afinado de maneira diferente que se partir de um Ré. Assim, instrumentos de teclado devem ser afinados novamente quanto uma música é tocada em um outro tom. Ao longo da história, surgiram algumas propostas mostrando caminhos alternativos para a afinação, e re-distribuindo o Coma de forma que ele fique acumulado em notas pouco usadas. Porém, especialmente com a necessidade de tocar o cravo em um repertório vasto sem ter grandes pausas, o processo histórico-cultural consolidou a afinação de igual temperamento.

A afinação de igual temperamento já era conhecida desde Vincenzo Galilei, mas ganhou grande força depois do surgimento das peças da série Cravo Bem Temperado de Bach. A afinação de igual temperamento distribui o Coma igualmente por todas as notas da escala, de forma que nenhuma soa especialmente desafinada. O custo disso é que todas as notas soam levemente desafinadas.

Na afinação de igual temperamento, a razão entre as frequências de duas notas consecutivas é igual a $\sqrt[12]{2}$, de forma que ao fim de 12 notas a frequência obtida será $f \times (\sqrt[12]{2})^{12} = f \times 2$.

In [15]:
frequencias_t = [] # Esta lista recebera todas as frequencias de uma escala

ft = referencia_inicial
while len(frequencias_t) < 12:
    frequencias_t.append(ft)
    ft *= 2**(1/12.)


frequencias_t.sort()
print frequencias_t
print ft

[440.0, 466.1637615180899, 493.8833012561241, 523.2511306011974, 554.3652619537443, 587.3295358348153, 622.253967444162, 659.2551138257401, 698.456462866008, 739.988845423269, 783.9908719634989, 830.6093951598906]
880.0


### Quão desafinado é um sistema de afinação?

No experimento a seguir, vamos calcular o quão desafinado é cada sistema de afinação em relação aos [intervalos da escala diatônica](http://ray.tomes.biz/alex.htm). É importante lembrar que esses intervalos foram definidos ao longo de um processo histórico, e não através do resultado de um cálculo. Utilizaremos os seguintes:

| Intervalo | Razão de Frequências |
|:-----------:|:------------:|
| II       |       9/8 |  
| III     |      5/4 |  
| IV       |        4/3 |  
| V         |         3/2 |  
| VI       |       5/3 |  
| VII    |     15/8 |  



In [16]:
intervalos_diatonica = [2, 3, 4, 5, 6, 7]
intervalos_cromatica = [2, 4, 5, 7, 9, 11]
razoes = [9/8., 5/4., 4/3., 3/2., 5/3., 15/8.]

for i in xrange(len(intervalos_diatonica)):
    frequencia_ideal = referencia_inicial * razoes[i]
    frequencia_pitagorica = frequencias[intervalos_cromatica[i]]
    frequencia_temperada = frequencias_t[intervalos_cromatica[i]]
    
    erro_pitagorica = 100*(frequencia_pitagorica - (frequencia_ideal)) / (frequencia_ideal)
    erro_temperada = 100*(frequencia_temperada - (frequencia_ideal)) / (frequencia_ideal)
    
    print "Intervalo:", intervalos_diatonica[i]
    print "Erro pitagorica:", erro_pitagorica
    print "Erro temperada:", erro_temperada
    

Intervalo: 2
Erro pitagorica: 0.0
Erro temperada: -0.225595705834
Intervalo: 3
Erro pitagorica: 1.25
Erro temperada: 0.79368399159
Intervalo: 4
Erro pitagorica: 1.36432647705
Erro temperada: 0.112989062753
Intervalo: 5
Erro pitagorica: 0.0
Erro temperada: -0.112861541555
Intervalo: 6
Erro pitagorica: 1.25
Erro temperada: 0.907569830446
Intervalo: 7
Erro pitagorica: 1.25
Erro temperada: 0.679926686047


### Número MIDI
Em termos computacionals, rapidamente torna-se pouco prático utilizar rótulos como "Dó" ou "Ré" para se referir a notas musicais. No contexto da computação musical, é comum atribuir um número inteiro a cada nota, de forma que um intervalo de um semi-tom é obtido simplesmente somando ou subtraindo 1 do número relacionado (isto é, os números relacionados ao fá e ao mi, em uma mesma oitava, diferem de 1). Isso se tornou ainda mais importante com a popularização dos instrumentos musicais no padrão MIDI, em que se consolidou que uma nota de frequência fundamental $f$ deve ser associada a um inteiro $N$ calculado por:
$$N = 69 + 12 \log_2(f/440).$$

Isso significa que a frequência da nota $N$ é dada por:
$$f = 2^{(N-69)/12} \times 440.$$

Veja que este padrão associa o lá central (440 Hz) à nota de número 69. Esta escolha é arbitrária, mas acabou sendo consolidada ao longo do tempo.

A notação MIDI pode ser utilizada para associar alguns intervalos a deslocamentos na escala:

| Intervalo | Passos (MIDI) |
|:-----------:|:------------:|
| II- (semi-tom) | 1 |
| II (tom)      | 2 |  
| III-     |  3 | 
| III (terça)    |  4 | 
| IV  (quarta)     |        5 |  
| V- (trítono)    |  6 | 
| V  (quinta)       |         7 |  
| VI-       |       8 |  
| VI  (sexta)  |     9 |  
| VII-    |     10 |  
| VII  (sétima)  |     11 |
| VIII  (oitava)  |     12 |

### Notação Piano-Roll
A notação pianola, ou Piano-Roll, denota cada nota através de três valores: seu tempo de início, seu tempo de final e o número MIDI correspondente à sua altura. Esse tipo de notação pode ser mostrado graficamente, através de caixas cuja posição vertical representa a frequência (altura) e os limites horizontais representam o tempo em que a nota está ativa. Assim, obtemos algo parecido com isto:

![piano_roll](./mixcraft-screenshot.png)




## Exercícios

1. Faça um desenho representando uma notação piano-roll contendo as quatro primeiras notas utilizadas na afinação pitagórica partindo do lá central (A4). Suponha que as notas são tocadas por um segundo cada, em sequência (nunca simultâneas).
1. Crie um programa que simula um processo de afinação pitagórica, mas que usa intervalos de terça no lugar de intervalos de quinta. Verifique como, nesse sistema de afinação, se comporta o erro de afinação em relação aos intervalos ideais.
1. Encontre a diferença (percentual) entre as frequências de duas notas consecutivas na escala de igual temperamento.
1. Encontre a frequência fundamental, o número MIDI e o rótulo (no formato nota-oitava - A4, C#3, etc.) relacionados às notas correspondentes (se precisar, faça uma busca na Internet!):
 1. Às cordas de um violão, de um contrabaixo e de um violino;
 1. Às notas mínima e máxima típicas de sopranos, contraltos, tenores e barítonos;
 1. Às notas mais baixa e mais alta disponíveis num piano;
 1. Às notas de frequência mais baixa e mais alta que poderiam ser ouvidas por um ser humano.


